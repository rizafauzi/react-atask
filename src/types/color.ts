export interface IColor {
  primary50: string;
  primary100: string;
  primary300: string;
  primary500: string;
  text50: string;
  text100: string;
  text300: string;
  text500: string;
  line100: string;
  line200: string;
  white: string;
  black: string;
  bgBase: string;
  error500: string;
}

export type ColorType =
  | 'primary50'
  | 'primary100'
  | 'primary300'
  | 'primary500'
  | 'text100'
  | 'text300'
  | 'text500'
  | 'text50'
  | 'line100'
  | 'line200'
  | 'black'
  | 'white'
  | 'error500';
