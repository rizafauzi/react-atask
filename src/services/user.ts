import { IUser } from '@/types/user';
import { octokit } from '@configs/octokit';
import { useQuery } from 'react-query';

const fetchSearchUser = async (keyword: string) => {
  const response = await octokit.request('GET /search/users', {
    q: keyword,
    per_page: 5,
  });

  return response.data.items as IUser[];
};

export const useSearchUsers = (keyword: string) => {
  return useQuery(['users', keyword], () => fetchSearchUser(keyword), {
    enabled: !!keyword,
  });
};
