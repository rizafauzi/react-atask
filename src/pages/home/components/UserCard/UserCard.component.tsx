import { useCallback, useState } from 'react';
import { styled } from 'styled-components';

import { IUser } from '@/types/user';
import { Image } from '@components';
import { color } from '@styles/color';
import { ReactComponent as ArrowIcon } from '@assets/icons/arrow.icon.svg';

import RepoList from '../RepoList';

interface UserCardProps {
  data: IUser;
  className?: string;
}

const UserCard = ({ className, data }: UserCardProps) => {
  const { login, avatar_url } = data;
  const [isShowRepo, setIsShowRepo] = useState(false);

  const toggleShowRepo = useCallback(() => setIsShowRepo(prev => !prev), []);

  return (
    <UserCardWrapper data-testid="user-card" className={className}>
      <button className="label" onClick={toggleShowRepo}>
        <div className="avatar">
          <Image
            width={30}
            alt={login}
            height={30}
            quality={0.5}
            src={avatar_url}
            className="rounded-full"
          />
          <h4>{login}</h4>
        </div>
        <ArrowIcon
          width={20}
          height={20}
          color="black"
          className={isShowRepo ? 'arrow-expand' : 'arrow-collapse'}
        />
      </button>
      {isShowRepo && (
        <div className="repo-content">
          <RepoList name={login} />
        </div>
      )}
    </UserCardWrapper>
  );
};

const UserCardWrapper = styled.div`
  .label {
    width: 100%;
    display: flex;
    border-radius: 8px;
    padding: 12px 18px;
    align-items: center;
    transition: all 0.2s ease-out;
    justify-content: space-between;
    outline: solid 0px transparent;
    border: solid 0.5px ${color.line200};

    .arrow-expand {
      transition: all 0.2s ease-out;
      transform: rotate(180deg);
    }

    .arrow-collapse {
      transition: all 0.2s ease-out;
      transform: rotate(0deg);
    }

    .avatar {
      gap: 12px;
      display: flex;
      flex-direction: row;
      align-items: center;
    }

    &:hover {
      cursor: pointer;
      background: ${color.primary50};
      outline: solid 1px ${color.primary300};
    }
  }

  .repo-content {
    padding-top: 8px;
    padding-left: 24px;
  }
`;

export default UserCard;
