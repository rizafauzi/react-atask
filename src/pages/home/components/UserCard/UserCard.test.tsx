import { cleanup, render, screen } from '@testing-library/react';

import { IUser } from '@/types/user';

import UserCard from './UserCard.component';

const data: IUser = {
  login: 'rizafauzi',
  id: 46235432,
  node_id: 'MDQ6VXNlcjQ2MjM1NDMy',
  avatar_url: 'https://avatars.githubusercontent.com/u/46235432?v=4',
  gravatar_id: '',
  url: 'https://api.github.com/users/rizafauzi',
  html_url: 'https://github.com/rizafauzi',
  followers_url: 'https://api.github.com/users/rizafauzi/followers',
  following_url:
    'https://api.github.com/users/rizafauzi/following{/other_user}',
  gists_url: 'https://api.github.com/users/rizafauzi/gists{/gist_id}',
  starred_url: 'https://api.github.com/users/rizafauzi/starred{/owner}{/repo}',
  subscriptions_url: 'https://api.github.com/users/rizafauzi/subscriptions',
  organizations_url: 'https://api.github.com/users/rizafauzi/orgs',
  repos_url: 'https://api.github.com/users/rizafauzi/repos',
  events_url: 'https://api.github.com/users/rizafauzi/events{/privacy}',
  received_events_url: 'https://api.github.com/users/rizafauzi/received_events',
  type: 'User',
  site_admin: false,
  score: 1,
};

afterEach(() => {
  cleanup();
});

test('should render UserCard component', () => {
  render(<UserCard key={data.id} data={data} />);
  const buttonElement = screen.getByTestId('user-card');
  expect(buttonElement).toBeInTheDocument();
  expect(buttonElement).toHaveTextContent(data.login);
});

test('matches snapshot', () => {
  const userCard = <UserCard key={data.id} data={data} />;
  expect(userCard).toMatchSnapshot();
});
