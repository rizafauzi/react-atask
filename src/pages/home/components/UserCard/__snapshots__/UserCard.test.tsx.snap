// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<UserCard
  data={
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/46235432?v=4",
      "events_url": "https://api.github.com/users/rizafauzi/events{/privacy}",
      "followers_url": "https://api.github.com/users/rizafauzi/followers",
      "following_url": "https://api.github.com/users/rizafauzi/following{/other_user}",
      "gists_url": "https://api.github.com/users/rizafauzi/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/rizafauzi",
      "id": 46235432,
      "login": "rizafauzi",
      "node_id": "MDQ6VXNlcjQ2MjM1NDMy",
      "organizations_url": "https://api.github.com/users/rizafauzi/orgs",
      "received_events_url": "https://api.github.com/users/rizafauzi/received_events",
      "repos_url": "https://api.github.com/users/rizafauzi/repos",
      "score": 1,
      "site_admin": false,
      "starred_url": "https://api.github.com/users/rizafauzi/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/rizafauzi/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/rizafauzi",
    }
  }
/>
`;
